bundle/src/main/java/com/adobe/acs/commons/wcm/impl/RobotsServlet.java

```diff
@@ -0,0 +1,418 @@
+        if (printGroupingComments) {
+            writer.println("# Start Group: " + group.getGroupName());
+        }
+
+        PageManager pageManager = request.getResourceResolver().adaptTo(PageManager.class);
+        Page page = pageManager.getContainingPage(request.getResource());
+
+        group.getUserAgents().stream().map(this::buildUserAgentsString).forEach(writer::println);
+
+        group.getAllowed().stream().map(allowed -> buildAllowedString(allowed, request.getResourceResolver())).forEach(writer::println);
+
+        List<String> allowProperties = group.getAllowProperties();
+        if (!allowProperties.isEmpty()) {
+            addRulesFromPages(page, request.getResourceResolver(), allowProperties, writer, this::buildAllowedString);
+        }
+
+        group.getDisallowed().stream().map(disallowed -> buildDisallowedString(disallowed, request.getResourceResolver())).forEach(writer::println);
+
+        List<String> disallowProperties = group.getDisallowProperties();
+        if (!disallowProperties.isEmpty()) {
+            addRulesFromPages(page, request.getResourceResolver(), disallowProperties, writer, this::buildDisallowedString);
+        }
+
+        if (printGroupingComments) {
+            writer.println("# End Group: " + group.getGroupName());
+        }
+    }
+
+    private void addRulesFromPages(Page page, ResourceResolver resourceResolver, List<String> propNames, PrintWriter writer, BiFunction<Page, ResourceResolver, String> ruleBuilderFunc) {
+        ValueMap pageProps = page.getProperties();
+        boolean added = false;
+        for (String prop : propNames) {
+            boolean shouldAdd = pageProps.get(prop, false);
+            if (shouldAdd) {
+                String rule = ruleBuilderFunc.apply(page, resourceResolver);
+                writer.println(rule);
+                added = true;
+                break;
+            }
+        }
+
+        if (!added) {
+            //since the rules are added to (dis)allow a page and it's children, we only need to recurse if no rule is added for the current page.
+            Iterator<Page> pageIterator = page.listChildren(new PageFilter(false, true), false);
+            pageIterator.forEachRemaining(child -> addRulesFromPages(child, resourceResolver, propNames, writer, ruleBuilderFunc));
+        }
+
+    }
+
+    private void writeFromJcrProperty(SlingHttpServletRequest request, SlingHttpServletResponse response) throws IOException {
+        if (!robotsContentsPropertyPath.startsWith("/")) {
+            robotsContentsPropertyPath = request.getResource().getPath() + "/" + robotsContentsPropertyPath;
+        }
+
+        Session session = request.getResourceResolver().adaptTo(Session.class);
+        boolean written = false;
+        try {
+            if (session.itemExists(robotsContentsPropertyPath)) {
+                Item item = session.getItem(robotsContentsPropertyPath);
+                if (item instanceof Property) {
```

Comment:
Should we log a WARN if this points to something that isn't a property/string/binary?

Reply:
added some additional logging to make this less ambiguous.